feat: add multichain account intro modal

fabiobozzo opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR implements the educational modal for Multichain Accounts (State 2) as specified in MUL-416. The modal introduces users to the new multichain UI/UX and is required for privacy policy compliance.

**What is the reason for the change?**
As part of the multichain accounts rollout, we need to educate users about the new functionality when they upgrade to State 2. This ensures users understand how multichain accounts work before they start using them.

**What is the improvement/solution?**
- Created a modal that appears once when the State 2 feature flag is enabled
- Provides educational content about multichain accounts with placeholder graphics
- Includes a "View Accounts" button that triggers wallet alignment with a minimum 2-second loading state
- Navigates users to the account list after alignment completes
- Tracks user interactions via MetaMetrics
- Works in both popup and fullscreen extension modes
- Persists state to ensure the modal only shows once per user

## **Changelog**

CHANGELOG entry: Added educational modal for Multichain Accounts feature introduction

## **Related issues**

[MUL-416](https://consensyssoftware.atlassian.net/browse/MUL-416)

## **Manual testing steps**

1. Enable the State 2 multichain accounts feature flag
2. Open MetaMask extension (both popup and fullscreen modes)
3. Verify the modal appears on the main wallet view / balances (not during onboarding)
4. Click "View Accounts" and verify:
   - Loading state shows for minimum 2 seconds
   - Modal closes after loading completes
   - Navigation occurs to the account list page
5. Verify the modal does not appear again after being shown once
6. Test "Learn More" button opens support documentation (URL is TBD)
7. Test close button (X) dismisses modal and marks it as shown
8. Verify modal does not interrupt onboarding flow

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

https://github.com/user-attachments/assets/e8d33892-7699-4082-84ac-2f8dea1c32c0

### **Before**

- No educational modal for multichain accounts
- Users directly see multichain functionality without introduction

### **After**

- Educational modal appears when State 2 is enabled
- Users are guided through multichain accounts introduction
- Smooth transition to account list with loading feedback

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[MUL-416]: https://consensyssoftware.atlassian.net/browse/MUL-416?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
feat: add multichain account intro modal

- Implemented state management for tracking whether the multichain intro modal has been shown to the user.
- Added actions to align multichain wallets and check alignment status.
- Updated metrics to track user interactions with the new modal features.

Commit:
center button loading

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
pretty-lint

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
feat: add multichain account intro modal content and state management

- Introduced new messages for the multichain account intro modal in `messages.json`, including titles, descriptions, and button texts.
- Updated the state snapshot to track whether the multichain intro modal has been shown.
- Refactored the multichain account intro modal component to improve loading state management and ensure proper cleanup on unmount.
- Enhanced the modal container to manage the modal's visibility based on multichain account state.
- Adjusted routing logic to trigger the multichain intro modal when appropriate.

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
fix dialog look and feel

fabiobozzo commented on ui/components/app/modals/multichain-account-intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,94 @@
+    dispatch(hideModal());
+
+    // Navigate to account list
+    history.push(ACCOUNT_LIST_PAGE_ROUTE);
```
✅ State updates immediately through proper Redux dispatch
✅ Modal won't reappear due to state sync issues
✅ Cleaner, more predictable state management
✅ No error handling needed for synchronous Redux actions

Commit:
fix: MultichainAccountIntroModalContainer to use dispatch for setting modal visibility

- Updated the handleClose function to dispatch the action for setting the multichain intro modal as shown, improving state management.
- Removed the async/await pattern for setting the modal visibility, simplifying the code structure.

hmalik88 commented on ui/components/app/modals/multichain-account-intro-modal/multichain-account-intro-modal.component.tsx:
```diff
@@ -0,0 +1,195 @@
+        gap={3}
+        style={{ minHeight: '120px' }}
+      >
+        {/* Top row of circles */}
+        <Box display={Display.Flex} gap={3}>
+          <Box
+            style={{ width: '60px', height: '60px' }}
+            backgroundColor={BackgroundColor.errorMuted}
+            borderRadius={BorderRadius.full}
+          />
+          <Box
+            style={{ width: '60px', height: '60px' }}
+            backgroundColor={BackgroundColor.warningMuted}
+            borderRadius={BorderRadius.full}
+          />
+        </Box>
+
+        {/* Bottom row of circles */}
+        <Box display={Display.Flex} gap={3}>
+          <Box
+            style={{ width: '60px', height: '60px' }}
+            backgroundColor={BackgroundColor.successMuted}
+            borderRadius={BorderRadius.full}
+          />
+          <Box
+            style={{ width: '60px', height: '60px' }}
+            backgroundColor={BackgroundColor.infoMuted}
+            borderRadius={BorderRadius.full}
+          />
+        </Box>
```
Is this supposed to display something while we don't have a graphic to place here? I don't see anything in the screen recording.

fabiobozzo replied:
Indeed, the recording is stale, I'm going to reupload it asap, with the look&feel fixes as well as the graphic placeholder.

fabiobozzo replied:
@hmalik88 I updated the recording in PR description ✔️

hmalik88 commented on ui/components/app/modals/multichain-account-intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,89 @@
+      },
+    });
+
+    const startTime = Date.now();
+
+    try {
+      // Trigger multichain wallet alignment
+      await alignMultichainWallets();
+    } catch (error) {
+      console.error('Failed to align multichain wallets:', error);
+      // Continue even if alignment fails
+    }
+
+    // Ensure minimum 2 seconds loading time
+    const elapsedTime = Date.now() - startTime;
+    const remainingTime = Math.max(0, 2000 - elapsedTime);
+    if (remainingTime > 0) {
+      await new Promise((resolve) => setTimeout(resolve, remainingTime));
+    }
```
We don't need the catch because that method will never throw, we use `Promise.allSettled` at the group level and the error is logged already for each group.

hmalik88 replied:
This method is good! I guess I was trying to figure out how we got to the 2 second number haha. I talked about this a little bit in our slack channel. In the future maybe we can introduce specific timeouts for specific devices/connections :)

hmalik88 commented on ui/pages/routes/routes.component.tsx:
```diff
@@ -470,14 +489,61 @@ export default function Routes() {
+    const flagJustEnabled =
+      isMultichainAccountsState2Enabled &&
+      !prevMultichainProps.isMultichainAccountsState2Enabled;
+    const flagEnabledFirstTime =
+      isMultichainAccountsState2Enabled &&
+      isMainWalletArea &&
+      !hasShownMultichainIntroModal;
```
```suggestion
    const flagEnabledFirstTime =
      isMultichainAccountsState2Enabled &&
      isMainWalletArea;
```

I think that last condition is duplicative to the condition below

hmalik88 reviewed (commented):
In the screen recording, the contents of the modal don't look centered and neither does the loading spinner in the view accounts button. But not sure if that's stale because in the screenshot you shared with @gantunesr, things look centered.

gantunesr commented on shared/constants/metametrics.ts:
```diff
@@ -735,6 +735,8 @@ export enum MetaMetricsEventName {
   MetricsOptIn = 'Metrics Opt In',
   MetricsOptOut = 'Metrics Opt Out',
   MetricsDataDeletionRequest = 'Delete MetaMetrics Data Request Submitted',
+  MultichainAccountIntroViewAccounts = 'Multichain Account Intro View Accounts Clicked',
+  MultichainAccountIntroLearnMore = 'Multichain Account Intro Learn More Clicked',
```
Lets check with Thomas! Because if we add new events we also need to update the data schemas and add them to mobile too

fabiobozzo replied:
Cool, thank you! I reverted it, and will keep the idea for later. Any other metrics we could benefit from? 💡

Commit:
address pr remarks

fabiobozzo commented on ui/pages/routes/routes.component.tsx:
```diff
@@ -470,14 +489,61 @@ export default function Routes() {
+      isUnlocked &&
+      !hasShownMultichainIntroModal &&
+      isMainWalletArea &&
+      (flagJustEnabled || flagEnabledFirstTime);
```
I simplified it, thanks! I want to only show modal when the feature flag transitions from disabled to enabled (upgrade scenario only).

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
refactor: remove multichain account intro event tracking from modal

Commit:
feat: add hasShownMultichainIntroModal flag and update modal display logic

- Introduced a new state flag `hasShownMultichainIntroModal` to track if the multichain intro modal has been displayed to the user.
- Updated the conditions for showing the modal to include cases where the feature flag is enabled for the first time on a fresh install.
- Refactored modal import to use the container component for better modularity.

Commit:
refactor: update import for MultichainAccountIntroModal to use container component

- Changed the import statement for the MultichainAccountIntroModal to use the MultichainAccountIntroModalContainer for improved modularity.
- Updated the modal contents to reflect this change.

Commit:
modal shows up only on transition from state 1 to 2

Commit:
lint: correct state access for hasShownMultichainIntroModal

PatrykLucka commented on ui/components/app/modals/multichain-account-intro-modal/multichain-account-intro-modal.component.tsx:
```diff
@@ -0,0 +1,195 @@
+  const [isLoading, setIsLoading] = useState(false);
+  const isMountedRef = useRef(true);
+
+  const handleClose = () => {
```
If it only triggers another function, we don't need it, we can use `onClose` directly

Commit:
removed the `handleClose` and `handleLearnMore` functions, directly using `onClose` and `onLearnMore` props instead

Commit:
fix: ensure en_GB multichain intro messages visibility on GitHub

- All multichain intro messages are present in app/_locales/en_GB/messages.json
- This commit addresses @gantunesr feedback about missing en_GB localization
- GitHub display issue prevented visibility of existing commit 3c7b668270
- Temp file added to trigger new commit for PR visibility

Commit:
cleanup: remove temporary file used for GitHub visibility

- Removed temporary file that was only needed to trigger commit visibility on GitHub
- All multichain en_GB messages remain present and committed in previous commits

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
refactor: update modal display logic for multichain accounts

- Simplified conditions for showing the multichain intro modal to only trigger on the main wallet route.
- Added logic to display the modal for fresh installs where the feature flag is already enabled.

Commit:
fix: updated the conditions for displaying the multichain intro modal to only trigger on feature flag transitions from disabled to enabled.

- Removed logic for showing the modal on fresh installs where the flag is already enabled.
- Added detailed console logging for debugging modal display conditions.

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
updated default fixture and onboarding fixture to include `hasShownMultichainIntroModal` for consistent state management across tests

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
test: Add `hasShownMultichainIntroModal` to state snapshots for consistent test state management

fabiobozzo commented on ui/pages/routes/routes.component.tsx:
```diff
@@ -470,14 +489,55 @@ export default function Routes() {
+      !hasShownMultichainIntroModal &&
+      isMainWalletArea &&
+      flagJustEnabled;
+
```
@gantunesr  the isMainWalletArea logic looks correct - it restricts the modal to only the home route (DEFAULT_ROUTE = "/") as intended. It prevents the modal from appearing on other routes in popup mode. Do you think it's incorrect? 👀

Commit:
refactor: reorganize multichain account intro modal structure

- Updated import path for `MultichainAccountIntroModalContainer` to reflect new directory structure.
- Removed obsolete `multichain-account-intro-modal` component and container files, streamlining the modal implementation.
- Added `usedInUi` property to `hasShownMultichainIntroModal` in app state for improved state management.

Commit:
fix imports

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
refactor: update state selector for multichain intro modal

- Changed the state selector for `hasShownMultichainIntroModal` from `state.metamask` to `state.appState` for improved state management consistency.

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
feat: add hasShownMultichainIntroModal state property

- Introduced `hasShownMultichainIntroModal` to the AppStateController and updated relevant types.
- Adjusted the selector in the Routes component to reference the correct state path for multichain modal tracking.

gantunesr commented on app/_locales/en/messages.json:
```diff
@@ -3515,6 +3515,34 @@
+    "message": "Same address, more networks",
+    "description": "Title for the second section of the multichain account intro modal"
+  },
+  "multichainAccountIntroSettingUp": {
```
hey @fabiobozzo, it looks like this one is not being used

Commit:
chore: remove unused multichain account intro message keys from localization files

Commit:
feat: add loading text for multichain account setup

- Introduced a new message key for loading text during multichain account setup in localization files.
- Updated the MultichainAccountIntroModal component to display the loading text while accounts are being set up.
- Adjusted the modal container to manage loading state and ensure proper user feedback during the account alignment process.

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
prettify

Commit:
feat: implement multichain upgrade detection and modal display logic

- Added logic to detect version upgrades and conditionally show the multichain intro modal based on the user's last seen version and feature flag status.
- Introduced new state properties in AppStateController for managing modal visibility and last seen version.
- Updated relevant components and actions to support the new modal display logic, ensuring a seamless user experience during upgrades.

Commit:
prettier run

Commit:
fix tests and checks

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
fix tests and checks

PatrykLucka commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,83 @@
+      try {
+        await Promise.all([
+          alignMultichainWallets(),
+          new Promise((resolve) => setTimeout(resolve, 2000)), // Minimum 2s UX feedback
```
If I understood requirement correctly, the 2 sec timeout should be started on button click and here it's starting on the component render, so user might never see a loading state if they will click the button after alignement will finish and 2 sec pass

fabiobozzo replied:
The `Promise.all` with 2-second timeout is now inside `handleViewAccounts()` (the button click handler), not in `useEffect`. The loading starts exactly when the user clicks "View Accounts", as intended. Does that make sense @PatrykLucka ?

Commit:
refactor: update multichain intro modal logic and app state management

- Removed version-based upgrade detection for the multichain intro modal.
- Introduced logic to disable the modal for fresh installs upon onboarding completion.
- Updated default state for `shouldShowMultichainIntroModal` to true for existing users.
- Cleaned up related tests and state management to reflect these changes.

Commit:
test whats-new-modal approach

fabiobozzo commented on app/scripts/metamask-controller.js:
```diff
@@ -1819,6 +1849,11 @@ export default class MetamaskController extends EventEmitter {
+    // Initialize upgrade detection after everything is set up
+    setTimeout(() => {
+      this.initializeUpgradeDetection();
+    }, 2000);
```
That entire approach has been removed now! I switched to the simple `lastUpdatedAt` pattern (same as the existing whats-new update modal). No more complex timeouts or version detection.

Commit:
cleanup previous approach leftovers

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

Commit:
refactor: update multichain account intro modal implementation

- Added `isOpen` prop to `MultichainAccountIntroModal` for better state management.
- Replaced the previous modal display logic with a local state in the `Routes` component.
- Removed unused modal import and cleaned up related code.
- Enhanced app state controller to include `includeInStateLogs` for improved logging.

ccharly commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.component.tsx:
```diff
@@ -0,0 +1,179 @@
+        </Box>
+      </Box>
+
+      {/* Action buttons */}
+      <Box
+        display={Display.Flex}
+        flexDirection={FlexDirection.Column}
+        width={BlockSize.Full}
+        gap={3}
+        marginTop={4}
+      >
+        <Button
+          variant={ButtonVariant.Primary}
+          size={ButtonSize.Lg}
+          width={BlockSize.Full}
+          onClick={onViewAccounts}
+          loading={isLoading}
+          disabled={isLoading}
+        >
+          {isLoading
+            ? t('multichainAccountIntroSettingUp')
+            : t('multichainAccountIntroViewAccounts')}
+        </Button>
+
+        <Button
+          variant={ButtonVariant.Link}
+          size={ButtonSize.Lg}
+          width={BlockSize.Full}
+          onClick={onLearnMore}
+          disabled={isLoading}
+        >
+          {t('multichainAccountIntroLearnMore')}
+        </Button>
```
Should this be a in `ModalFooter` instead?

ccharly commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,74 @@
+    );
+  }, []);
+
+  const handleClose = useCallback(() => {
```
Just made some comment on the mobile PR for this too:
- https://github.com/MetaMask/metamask-mobile/pull/19970#discussion_r2367790497

And looks like we want to wait for the alignement to finalize before closing

fabiobozzo replied:
`handleClose` now awaits `alignmentPromiseRef.current` before proceeding, so alignment completes before closing the modal.

ccharly commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,74 @@
+  const history = useHistory();
+  const [isLoading, setIsLoading] = useState(false);
+
+  const handleViewAccounts = useCallback(async () => {
+    // Start loading when user clicks (not when modal opens)
+    setIsLoading(true);
+
+    try {
+      await Promise.all([
+        alignMultichainWallets(),
+        new Promise((resolve) => setTimeout(resolve, 2000)), // Minimum 2s UX feedback
+      ]);
+    } catch (error) {
+      console.error('Wallet alignment failed:', error);
+      // Even if alignment fails, continue to account list
+    }
+
+    // Mark modal as shown so it doesn't show again
+    dispatch(setMultichainIntroModalShown(true));
+    onClose();
+
+    // Navigate to account list
+    history.push(ACCOUNT_LIST_PAGE_ROUTE);
+  }, [dispatch, history, onClose]);
```
Could you use the same pattern than in mobile?
- https://github.com/MetaMask/metamask-mobile/pull/19970/files#diff-34690b1add0c5cae06b3aeb4961a268e8982d5503a8e83fe1e2d7a829daefa7bR37-R58
- https://github.com/MetaMask/metamask-mobile/pull/19970/files#diff-34690b1add0c5cae06b3aeb4961a268e8982d5503a8e83fe1e2d7a829daefa7bR84-R106

Basically:
- We want to capture those errors
- We use `useMemo` to keep a ref to the on-going promise
- We can safely `await` this promise in `handleViewAccounts`, the `useEffect` and potentially in the `handleClose` now too

fabiobozzo replied:
I implemented the mobile pattern with:
- `useMemo` to keep a ref to the on-going promise 
- capture errors in the promise handling
- await "safely" in handleViewAccounts and handleClose
- cleanup `useEffect` for component unmount 

It should now match the mobile implementation pretty much.

ccharly commented on ui/pages/routes/routes.component.tsx:
```diff
@@ -856,6 +905,13 @@ export default function Routes() {
       {isUnlocked ? <Alerts history={history} /> : null}
       <ToastMaster />
+
+      {showMultichainIntroModal ? (
+        <MultichainAccountIntroModalContainer
+          isOpen={showMultichainIntroModal}
+          onClose={() => setShowMultichainIntroModal(false)}
+        />
+      ) : null}
```
Any reason why this is not closer to the other modals?

Commit:
address pr feedback from ccharly

- Introduced `ModalFooter` to improve layout and organization of action buttons.
- Updated button handling in `MultichainAccountIntroModal` for better user experience.
- Implemented promise management in `MultichainAccountIntroModalContainer` to handle wallet alignment more effectively.
- Added error handling for wallet alignment process to provide user feedback.
- Cleaned up modal display logic in the `Routes` component for improved readability.

Commit:
fix: replaced individual button components in MultichainAccountIntroModal with a unified ModalFooter

- Enhanced button properties for better user experience during loading states.
- Updated promise handling in `MultichainAccountIntroModalContainer` to ensure alignment completion feedback is clear and consistent.

Commit:
fix footer buttons

Commit:
fix footer buttons

Commit:
fix footer buttons

Commit:
fix footer buttons

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

fabiobozzo commented:
Testing strategy for this feature:

1.
```
git checkout v13.3.0 # a previous release
_flags.remoteFeatureFlags.enableMultichainAccounts.enabled = false # override

- clean build
- onboarding with new srp
- no modal shows up
- close extension
```

2. (State-2 upgrade)
```
git checkout main 
_flags.remoteFeatureFlags.enableMultichainAccounts.enabled = true # override

- clean build
- unlock extension
- modal shows up ✨ 
```

3. (fresh install)
```
- remove extension
- install it again
- no modal shows up  👻 
```

Commit:
fix: resolve wallet alignment race condition in multichain intro modal

- Move alignment promise creation from useMemo to handleViewAccounts to fix race condition
- Fix TypeScript Promise<[void, void]> vs Promise<void> type mismatch
- Ensure wallet alignment actually runs when "View Accounts" is clicked
- Address cursor bot feedback on stale promise handling

Commit:
fix: clear loading state before modal closes to prevent React warnings

- Update test snapshot for hasShownMultichainIntroModal

Commit:
fix cursorbot remarks

Commit:
fix cursorbot remarks

Commit:
fix: address ccharly feedback

- Mobile alignment pattern (start on modal open)
- Add timeout/URL constants
- Rename to MultichainAccountsIntroModal
- Clean up comments and fix ESLint issues
- Update controller, actions, routes, and tests

fabiobozzo commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,117 @@
+    try {
+      // Create the alignment promise immediately when clicked
+      const alignmentPromise = Promise.all([
+        alignMultichainWallets(),
```
Removed by mistake by a previous refactoring: now again alignment starts on modal open. ✔️

Commit:
refactor: extract multichain intro modal logic into a custom hook for better organization

- Replace TODO comments with actual error reporting using Sentry
- Remove unused state and related logic for showing the intro modal

Commit:
Merge branch 'main' into feat/MUL-416-multichain-account-intro-modal

fabiobozzo commented on ui/components/app/modals/multichain-accounts/intro-modal/multichain-account-intro-modal.container.tsx:
```diff
@@ -0,0 +1,117 @@
+      // Wait for alignment to complete
+      await alignmentPromise;
+    } catch (err) {
+      console.error('Wallet alignment failed:', err);
```
Done, can you please check? 🙏

Commit:
fix lint and ci failures

Commit:
fix lint and ci failures

Commit:
fix lint and ci failures

Commit:
address pr remarks from ccharly

Commit:
address pr remarks from ccharly

ccharly reviewed (dismissed):
LGTM (code-wise), let's give it another round of testing and we should be ok! (still approving for now)

Commit:
prettify
